packages/compass-aggregations/src/components/pipeline-builder-workspace/pipeline-builder-ui-workspace.jsx

```diff
@@ -87,12 +90,31 @@ export class PipelineBuilderUIWorkspace extends PureComponent {
 const mapState = (state) => {
   return {
     stageIds: state.pipelineBuilder.stageEditor.stageIds,
-    editViewName: state.editViewName
-  }
-};
-
-const mapDispatch = {
-  onStageMoveEnd: moveStage
+    editViewName: state.editViewName,
+    stageOperators: state.pipelineBuilder.stageEditor.stages.map(
+      (stage) => stage.stageOperator
+    ),
+  };
 };
 
-export default connect(mapState, mapDispatch)(PipelineBuilderUIWorkspace);
+export default connect(
+  mapState,
+  { moveStage },
+  (stateProps, dispatchProps, ownProps) => {
+    const { stageOperators, ...restOfStageProps } = stateProps;
+    const { moveStage } = dispatchProps;
+    return {
+      ...ownProps,
+      ...restOfStageProps,
+      onStageMoveEnd: (fromIndex, toIndex) => {
+        track('Aggregation Edited', {
+          num_stages: stageOperators.length,
+          stage_action: 'stage_reordered',
+          stage_name: stageOperators[fromIndex],
+          editor_view_type: 'stage', // Its always stage view for this component
+        });
```

Comment:
Would it make sense to keep these inside the actions instead of connect method? Looking at the code you almost always have a tracking event in the connect method right near some action, and you always need to additionally add some data to `mapState` to then get it in `mergeProps` to produce an event, it looks like trying to simulate action creator through connect method, and so this kinda looks to me that it might be a bit cleaner to move it all in the action itself, especially as the actions are pretty UI-specific (with the only exception being `openPipelinebyId` usage in pipeline card where event was already tracked from inside UI)

Reply:
actually `openPipelinebyId` triggered this approach for adding events in `mergeProps`. I didn't want to pass in the `num_stages` to the component and then use it just for tracking. I'll be moving all others to action handlers.

Reply:
I don't see `num_stages` in the tracking event there though, do you mean view type or am I looking at something else? 🤔

Reply:
ah yeah! sorry my bad!